site/app/models/gradeable/Gradeable.php

```diff
@@ -0,0 +1,446 @@
+            $this->setPeerGradeSet($details["peer_grade_set"]);
+            $this->setLateSubmissionAllowed($details["late_submission_allowed"]);
+            $this->setPrecision($details["precision"]);
+        }
+
+        // Set dates last
+        $this->setDates($details);
+    }
+
+    const dates = [
+        'ta_view_start_date',
+        'grade_start_date',
+        'grade_released_date',
+        'team_lock_date',
+        'submission_open_date',
+        'submission_due_date',
+        'grade_locked_date'
+    ];
+    public function toArray()
+    {
+        // Use the default behavior for the most part, but convert the dates
+        $return = parent::toArray();
+
+        foreach(self::dates as $date) {
+            $return[$date] = $this->$date !== null ? DateUtils::dateTimeToString($this->$date) : null;
+        }
+
+        return $return;
+    }
+
+    private function loadAutogradingConfig() {
+        $course_path = $this->core->getConfig()->getCoursePath();
+        $details = FileUtils::readJsonFile(FileUtils::joinPaths($course_path, "config", "build",
+            "build_{$this->id}.json"));
+
+        if (isset($details['max_submission_size'])) {
+            $details['max_submission_size'] = floatval($details['max_submission_size']);
+        }
+
+        if (isset($details['max_submissions'])) {
+            $details['max_submissions'] = intval($details['max_submissions']);
+        }
+
+        if (isset($details['assignment_message'])) {
+            $details['assignment_message'] = Utils::prepareHtmlString($details['assignment_message']);
+        }
+
+        return $details;
+    }
+    public function getAutogradingConfig()
+    {
+        if($this->type !== GradeableType::ELECTRONIC_FILE) {
+            throw new \BadFunctionCallException("Cannot load autograding config for non-electronic file!");
+        }
+
+        // use JIT loading
+        if ($this->autograding_config === null) {
+            $this->autograding_config = $this->loadAutogradingConfig();
+        }
+        return $this->autograding_config;
+    }
+
+
+    /* Overridden setters with validation */
+    private function validateDates(array &$dates)
+    {
+        $errors = [];
+
+        //
+        // Parse all of the dates into DateTime's (no error if null)
+        //
+        foreach(self::dates as $date) {
+            if($dates[$date] === null) continue;
+            $result = DateUtils::assertDate($dates[$date], $this->core->getConfig()->getTimezone());
+            if ($result !== null) {
+                $errors[$date] = $result;
+            }
+        }
+
+        $ta_view_start_date = $dates['ta_view_start_date'];
+        $grade_start_date = $dates['grade_start_date'];
+        $grade_released_date = $dates['grade_released_date'];
+        $team_lock_date = $dates['team_lock_date'];
+        $submission_open_date = $dates['submission_open_date'];
+        $submission_due_date = $dates['submission_due_date'];
```

Comment:
These need to be by reference otherwise you're modifying nothing

Reply:
If you're referring to the assignment on lines 253 through 258, these are just alias's to get named variables into the scope instead of looking up a name in the `dates` array every time to reduce clutter.  After the `foreach`loop, none of the values should be modified.

Reply:
I assumed you wanted to modify the dates array as well because you were modifying `$late_days` and `$grade_start_date` later in the function

Reply:
Thanks for catching that, should be fixed now